File: 543. Diameter of Binary Tree-1.cpp
class Solution {
public:
  int res = 0;

  int diameterOfBinaryTree(TreeNode *root) {
    helper(root);

    return res;
  }
  int helper(TreeNode *root) {
    if (!root)
      return 0;

    int l = helper(root->left);
    int r = helper(root->right);

    res = max(res, l + r);

    return max(l, r) + 1;
  }
};

File: 543. Diameter of Binary Tree-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(root):
            if root is None:
                return 0
            l_len = dfs(root.left)
            r_len = dfs(root.right)
            nonlocal res
            res = max(res, l_len + r_len)
            return max(l_len, r_len) + 1

        dfs(root)

        return res
